x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/saml/SamlRealm.java

```diff
@@ -744,25 +785,69 @@ private static Tuple<AbstractReloadingMetadataResolver, Supplier<EntityDescripto
+                } else {
+                    logger.warn(
+                        "cannot load SAML metadata for [{}] from [{}]; SAML authentication for this realm will fail",
+                        entityId,
+                        sourceLocation
+                    );
+                    return new UnresolvedEntity(entityId, sourceLocation);
+                }
             }
             return descriptor;
         } catch (ResolverException e) {
             throw SamlUtils.samlException("Cannot resolve entity metadata", e);
         }
     }
 
+    private static EntityDescriptor resolveEntityDescriptorWithPossibleRefresh(AbstractReloadingMetadataResolver resolver, String entityId)
+        throws ResolverException {
+        var criteria = new CriteriaSet(new EntityIdCriterion(entityId));
+        EntityDescriptor descriptor = resolver.resolveSingle(criteria);
+        if (descriptor == null) {
+            /*
+             * If the descriptor is null, we haven't found a metadata file with that entity id in it. This could be caused by 2 things:
+             * 1. We didn't find any metadata file (e.g. the metadata is loaded over http, and the request returned an error
+             * 2. The file does exist, but doesn't contain the entity.
+             * In either case it's worth refreshing the metadata again to see if it's now correct because it's possible that the problem
+             *    has been resolved (although if metadata is loaded from a local file we monitor it for changes anyway, so this refresh
+             *    is unlikely to have any benefit).
+             * The resolver's refresh method is synchronized, and could be a choke point if there are a lot of concurrent login attempts,
+             *    so we wrap it in a check to only call it on 1 thread. The other threads will fail due to bad metadata, but that's OK
+             *    because the metadata is (was) bad, and if the refresh works, then their next login attempt will also work.
+             * Since the dynamic refresh is a saving mechanism (we have a separate regular poll for refreshes) we prefer it to be safe
+             *    (from a thread contention point of view) rather than perfect (from a user experience point of view).
+             */
+            if (REFRESHING_METADATA.compareAndSet(false, true)) {
+                try {
+                    resolver.refresh();
+                } catch (ResolverException e) {
+                    // Refresh failed. Since there could be a lot of these and it's only a saving mechanism we log as debug
+                    // The resolver itself will log a WARN with the HTTP status etc.
+                    logger.debug(() -> "Failed to refresh SAML metadata for [" + entityId + "]", e);
+                } finally {
+                    REFRESHING_METADATA.compareAndSet(true, false);
+                }
+            }
+            descriptor = resolver.resolveSingle(criteria);
```

Comment:
Is it intentional that `resolver.resolveSingle` is called again even when the above `if` check returns false, i.e. when the thread is not performing the actual refresh?

Reply:
It is intentional, on the off chance that whichever other thread performed the refresh has completed in time. It will likely fail, but sometimes it will work.